src/state/slices/time_slice.ts

```diff
@@ -48,17 +57,24 @@ export const createTimeSlice: StateCreator<TimeSlice & DatasetSlice, [], [], Tim
     () => get().currentFrame,
     async (frame) => {
       set({ pendingFrame: frame });
-      await get().loadFrameCallback(frame);
-      set({ currentFrame: frame });
+      const result = await get().frameLoadCallback(frame);
+      if (result !== null) {
+        set({ currentFrame: result.frame, frameLoadResult: result });
+      } else if (get().pendingFrame === frame) {
+        // Reset pendingFrame if it hasn't changed. (e.g. no other calls to
+        // setFrame were made while this one was loading)
+        // TODO: More robust handling for requests? Request IDs?
+        set({ pendingFrame: get().currentFrame });
+      }
```

Comment:
Now that `frameLoadCallback` can report the status of the load, we can do more validation during this step! A `null` result means that the frame requested was out of bounds.